Mark all cuco kernels as static so they have hidden visibility

robertmaynard opened the pull request:
This marks all kernels in CUCO as `static` so that they have internal linkage and won't conflict when used by multiple DSOs.

I didn't see a single shared/common header in cuco where I could place a `CUCO_KERNEL` macro so I modified each instance instead.
While `cccl` went with a `__attribute__ ((visibility ("hidden")))` approach to help reduce RDC size, this approach seemed very invasive for cuco. This is due to the fact that we would need to pragma push and pop both gcc warnings and nvcc warnings in each cuco header so that we don't introduce any warnings. This is needed as the compiler incorrectly state that the `__attribute__ ((visibility ("hidden")))` has no side-effect.

Context:
https://github.com/rapidsai/cudf/pull/14726
https://github.com/NVIDIA/cccl/issues/166
https://github.com/rapidsai/raft/issues/1722

robertmaynard committed:
Mark all cuco kernels as static so they have hidden visibility

PointKernel commented:
/ok to test

PointKernel commented:
> I didn't see a single shared/common header in cuco where I could place a CUCO_KERNEL macro so I modified each instance instead.

https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/detail/utility/cuda.hpp probably the best place for such a macro

robertmaynard commented:
> > I didn't see a single shared/common header in cuco where I could place a CUCO_KERNEL macro so I modified each instance instead.
> 
> https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/detail/utility/cuda.hpp probably the best place for such a macro

Updated to introduce a CUCO_KERNEL macro to `cuda.hpp`.

jrhemstad reviewed (commented):
Since cuco is header-only and may be included in projects that are building with `rdc=true`, I think it is more important to preserve the `__attribute__ ((visibility ("hidden"))` functionality when `__CUDA_RDC__` is defined. 

As you mentioned, you will need to universally silence the `-Wattributes` warning as seen here: https://github.com/NVIDIA/cccl/blob/2771c61545eff4ec3cede24ba7963c4eebc4bbaf/cub/cub/util_macro.cuh#L128-L140

robertmaynard commented:
The issue I have with CCCL approach is that it has side effects on all functions inside the TU and not just the CCCL functions since the suppressions aren't push/popped. 

This could hide real issues in user code, and why I don't recommend universally silencing the warnings. But I will implement what ever approach the CuCo team wants, which by the sound of it is universal suppression

sleeepyjack commented:
I'm trying to wrap my head around this. So `static` is still correct, but we have an increased binary size in case `rdc=true`?

Using the `hidden` attribute would also help with the binary size, but adding pragma push/pop to every header (does it need to be _every_ header or just those that directly emit the warning?) is very invasive. Can we prefix the `CUCO_KERNEL` macro with `#pragma warning ( suppress: FooWarning)` instead?

robertmaynard commented:
> I'm trying to wrap my head around this. So `static` is still correct, but we have an increased binary size in case `rdc=true`?
> 
Correct.

> Using the `hidden` attribute would also help with the binary size, but adding pragma push/pop to every header (does it need to be _every_ header or just those that directly emit the warning?) is very invasive. Can we prefix the `CUCO_KERNEL` macro with `#pragma warning ( suppress: FooWarning)` instead?

It would help with binary size only in case of `rdc=true`. You are correct it will only need to be the headers that have kernels.


> Can we prefix the `CUCO_KERNEL` macro with `#pragma warning ( suppress: FooWarning)` instead?
You can but that will effect all following code, which would include user code inside any translation unit that included the header.

sleeepyjack commented:
> You can but that will effect all following code, which would include user code inside any translation unit that included the header.

Ah, ok, I thought `#pragma warning ( suppress:` would only affect the line that directly follows, not the entire TU.

I would prefer the `hidden` annotation solution over `static`, but if it's too much work I'm fine with the latter too.

Most of the kernels in cuco are already isolated in separate header files, so we would need to add the push/pop pragmas to 8 files in total (ignoring any kernels in test/benchmark directories):

- include/cuco/detail/open_addressing/kernels.cuh
- include/cuco/detail/static_map/kernels.cuh
- include/cuco/detail/static_multimap/kernels.cuh
- include/cuco/detail/static_set/kernels.cuh
- include/cuco/detail/trie/dynamic_bitset/kernels.cuh
- include/cuco/detail/storage/kernels.cuh
- include/cuco/detail/dynamic_map_kernels.cuh
- include/cuco/detail/static_map_kernels.cuh (legacy map impl)

robertmaynard commented:
> I would prefer the hidden annotation solution over static

I will go forward with that approach 👍

robertmaynard commented:
@sleeepyjack 
After testing with the cuco examples and libcudf, we can't pop the macros as desired.
This due to how the `__attribute__ ((visibility ("hidden")))` is applied during nvcc internal code generation. Consuming TUs will also generate attribute warnings and those can only be captured by not popping the pragmas.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

PointKernel commented:
/ok to test

sleeepyjack commented:
> Consuming TUs will also generate attribute warnings and those can only be captured by not popping the pragmas.

So this means we globally disable this warning in user land? Not ideal. @jrhemstad any ideas?

jrhemstad commented:
> So this means we globally disable this warning in user land? Not ideal. @jrhemstad any ideas?

That's precisely what we did in every CCCL header. It's a pretty innocuous warning to silence, and the benefits far outweigh any disadvantages.

PointKernel commented on examples/static_set/device_subsets_example.cu:
```diff
@@ -77,7 +77,7 @@ key_type constexpr empty_key_sentinel = -1;
  * @param set_refs Pointer to the array of subset objects
  */
-__global__ void insert(ref_type* set_refs)
+static __global__ void insert(ref_type* set_refs)
```
```suggestion
__global__ void insert(ref_type* set_refs)
```

PointKernel commented on examples/static_set/device_subsets_example.cu:
```diff
@@ -102,7 +102,7 @@ __global__ void insert(ref_type* set_refs)
  * @param set_refs Pointer to the array of subset objects
  */
-__global__ void find(ref_type* set_refs)
+static __global__ void find(ref_type* set_refs)
```
```suggestion
__global__ void find(ref_type* set_refs)
```

PointKernel reviewed (commented):
Checked with @sleeepyjack earlier today, this PR is ready to ship once we revert changes in the example. Examples are considered as user code thus no to need change it.

robertmaynard committed:
Update examples/static_set/device_subsets_example.cu

Co-authored-by: Yunsong Wang <yunsongw@nvidia.com>

robertmaynard committed:
Update examples/static_set/device_subsets_example.cu

Co-authored-by: Yunsong Wang <yunsongw@nvidia.com>

robertmaynard commented:
> Checked with @sleeepyjack earlier today, this PR is ready to ship once we revert changes in the example. Examples are considered as user code thus no to need change it.

I have integrated the changes to examples. So you can merge whenever you are ready.
Thanks!

sleeepyjack commented:
/ok to test
